qs web-components

ia3andy opened the pull request.

yrodiere commented on src/main/java/io/quarkus/search/app/hibernate/InputProviderHtmlBodyTextBridge.java:
```diff
@@ -18,7 +18,7 @@ public String toIndexedValue(InputProvider provider, ValueBridgeToIndexedValueCo
             Element body = Jsoup.parse(in, StandardCharsets.UTF_8.name(), "/").body();
             // Content div has two grid columns: actual content and TOC. There's not much use of the TOC, we want the content only:
-            Element content = body.selectFirst(".content .grid__item");
+            Element content = body.selectFirst(".content .grid_item");
```
I don't think that's right? At least it causes many problems when indexing:

```
2023-11-22 09:02:12,305 WARN  [io.qua.sea.app.hib.InputProviderHtmlBodyTextBridge] (executor-thread-1) Was unable to find the content section of a guide. Using whole document as text. GitInputProvider{path='version/2.13/guides/all-builditems.html'}
```

If I revert this change, the warnings disappear.

ia3andy replied:
ah I think this is IntelliJ refactor good catch!!!!!!

yrodiere commented on .gitignore:
```diff
@@ -5,6 +5,7 @@ pom.xml.releaseBackup
 pom.xml.versionsBackup
 release.properties
 .flattened-pom.xml
+node_modules/
```
Maybe a comment to explain the hack?

ia3andy replied:
ok

yrodiere commented on src/main/resources/web/app/qs-form.ts:
```diff
@@ -0,0 +1,174 @@
+    this.dispatchEvent(new CustomEvent(QS_START_EVENT));
+    const data = this._readQueryInputs();
+    this._loading = true;
+    this._jsonFetch(new AbortController(), 'GET', data, 1000)
+      .then((r: any) => {
+        let event = QS_RESULT_EVENT
+        if (this._page > 0 && r.hits.length > 0) {
+          this._total += r.hits.length;
+          event = QS_RESULT_NEXT_EVENT;
+        } else {
+          this._total = r.hits.length;
+        }
+        const hasMoreHits = r.total > this._total;
+
+        this.dispatchEvent(new CustomEvent(event, {detail: {...r, hasMoreHits}}));
+      }).finally(() => {
+      this._loading = false;
+      this.dispatchEvent(new CustomEvent(QS_END_EVENT));
+    });
```
I don't see offline (javascript) search handling in case we reach the timeout... Is it supposed to be implemented elsewhere?

ia3andy replied:
Yes, I need to see if we pass the data with json or if we need to parse the page content and send an event for this.

yrodiere reviewed (commented):
Thanks. And thanks for trying to keep some of my code, that's kind of you.

I added a few comments, hopefully some of them are useful :)

ia3andy commented on src/main/java/io/quarkus/search/app/dto/SearchHit.java:
```diff
@@ -5,5 +5,6 @@
 @ProjectionConstructor
-public record SearchHit(@IdProjection String id, @FieldProjection String title) {
+public record SearchHit(@IdProjection String id, @FieldProjection String title, @FieldProjection String summary,
+        @FieldProjection String keywords) {
```
Ok that was just an attempt, I will remove it.

ia3andy commented on src/main/resources/web/app/qs-form.ts:
```diff
@@ -0,0 +1,174 @@
+  private _handleInputChange = (e: Event) => {
+    const target = e.target as HTMLFormElement;
+    console.log(target.value);
+    if (this._isInput(target)) {
+      if (target.value.length === 0) {
+        this._clearSearch();
+        return;
+      }
+      if (target.value.length < this.minChars) {
+        return;
+      }
+    }
+    this._searchDebounced();
```
Yeah I need to have a look at the whole abort to implement it, I've ignored this and the timeout for now

ia3andy commented on src/main/resources/web/app/qs-target.ts:
```diff
@@ -0,0 +1,128 @@
+    if (this._result?.hits) {
+      if (this._result.hits.length === 0) {
+        return html`
+          <div id="qs-target">
+            <p>No results found</p>
+          </div>
```
For sure! same, I took shortcut to have something to share and iterate

ia3andy replied:
Fixed, need a bit more styling when integrating probably

yrodiere reviewed (commented):
Thanks, looking great.

A few comments below :)

ia3andy commented on src/main/resources/web/app/assets/icons/docsicon-concepts.svg:
They are configured to be bundled as base64 url, they won't be available directly.

yrodiere replied:
We still need copyright mentions for assets we include in our source... All the more so if we distribute them at runtime (however indirectly).

ia3andy commented on src/main/java/io/quarkus/search/app/WebComponents.java:
```diff
@@ -0,0 +1,35 @@
+    @Context
+    UriInfo uriInfo;
+
+    // This route allows to access the web-component js on a fixed path
+    // without affecting caching of the script
+    @Path("/search.js")
+    @Cache(noCache = true)
```
As discussed on Zulip, only the redirect will not be cached (the script will be cached), so that if the script changes, the the redirect will change (and be called again).

Also as discussed there is a price for this but I am not sure there is a way to know if it has changed without any call?

yrodiere replied:
> As discussed on Zulip, only the redirect will not be cached (the script will be cached), so that if the script changes, the the redirect will change (and be called again).

Seems fair.

> Also as discussed there is a price (on REST call) for this but I am not sure there is a way to know if it has changed without any call?

Makes sense.

ia3andy commented on src/main/resources/web/app/qs-guide.ts:
```diff
@@ -0,0 +1,124 @@
+    if(Array.isArray(content)) {
+      return content.map((c) => this._renderContent(c));
+    }
+    const parts = content.split(/___([^_]+)___/g);
```
I think it's pretty unlikely, but could we maybe have a detector before hand for it? 

It's nice to have the content human readable. I could settle on something more readable :-)

`_H__myword__L_` this one is very very unlikely even more with opening and closing tags

yrodiere replied:
> could we maybe have a detector before hand for it?

Not on the server side. Like I said the other day, there's HTML escaping, or nothing.

> It's nice to have the content human readable. I could settle on something more readable :-)

Yeah that's why we were using HTML ;D

> `_H__myword__L_` this one is very very unlikely even more with opening and closing tags

Ok, that seems safer.

ia3andy commented on src/main/resources/web/app/qs-guide.ts:
```diff
@@ -0,0 +1,124 @@
+
+  render() {
+    return html`
+      <div class="qs-hit qs-guide type-${this.data.type}">
+        <h4><a href="${this.data.url}">${this._renderContent(this.data.title)}</a></h4>
+        <div class="summary">
+          <p>${this._renderContent(this.data.summary)}</p>
+        </div>
+        <div class="keywords">${this._renderContent(this.data.keywords)}</div>
```
Ok I will add it!

ia3andy commented on src/main/resources/web/app/qs-guide.ts:
```diff
@@ -0,0 +1,124 @@
+    `;
+  }
+
+  private _renderContent(content: string) {
```
I'll give it a try

ia3andy commented:
Thanks @yrodiere for the review!!!

ia3andy commented:
Ok, I've pushed again:
- switched to html (and reverted the service changes)
- added a loading indicator
- added the content section

For the icons copyright, I've used the one from quarkusio so we should mimic what they are doing?

yrodiere commented:
> For the icons copyright, I've used the one from quarkusio so we should mimic what they are doing?

Yep, makes sense

ia3andy commented:
> Hey @ia3andy I see you marked this as ready for review, but just to make sure we agree: you intend to make some more changes, right? Regarding the copyright mentions in particular, and maybe the abort thing (see earlier reviews).
> 
> I do understand javascript search will be handled elsewhere.

Yes sorry I am stuck on code.quarkus, I will get back to this very soon

yrodiere commented:
> > Hey @ia3andy I see you marked this as ready for review, but just to make sure we agree: you intend to make some more changes, right? Regarding the copyright mentions in particular, and maybe the abort thing (see earlier reviews).
> > I do understand javascript search will be handled elsewhere.
> 
> Yes sorry I am stuck on code.quarkus, I will get back to this very soon

No problem, I just wanted to be sure there wasn't a misunderstanding. Take your time :+1:

yrodiere commented:
Hey @ia3andy , since you seem busy, and there are still a few things to fix here, and I think there's still a bit of work to get this integrated with quarkus.io, I'll plan on deploying the first version without web-components. We can always add them later when this is ready, of course.

marko-bekhta commented on pom.xml:
```diff
@@ -335,4 +367,67 @@
   </build>
+  <profiles>
+    <profile>
+      <id>locker</id>
```
It would be nice to have a comment on what this profile is for and when to use it, if ever 😃

marko-bekhta commented on pom.xml:
```diff
@@ -107,6 +107,31 @@
       <groupId>io.quarkus</groupId>
       <artifactId>quarkus-resteasy-reactive</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.quarkiverse.web-bundler</groupId>
+      <artifactId>quarkus-web-bundler</artifactId>
+      <version>${version.quarkus-web-bundler}</version>
+    </dependency>
+    <!-- Web Dependencies -->
+    <dependency>
+      <groupId>org.mvnpm</groupId>
+      <artifactId>lit</artifactId>
+      <version>3.1.2</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mvnpm</groupId>
+      <artifactId>lodash</artifactId>
+      <version>4.17.21</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mvnpm.at.types</groupId>
+      <artifactId>lodash</artifactId>
+      <version>4.14.202</version>
+      <scope>provided</scope>
+      <classifier>sources</classifier>
+    </dependency>
```
for other dependencies, we've tried to keep the versions under the dependency management section, and here, just use them without versions.

marko-bekhta commented on src/main/resources/web/app/qs-guide.ts:
```diff
@@ -0,0 +1,163 @@
+      font-weight: 400;
+    }
+
+    .qs-guide .content-highlights {
+      font-size: 0.7rem;
+    }
```
For this section where we are displaying the "full-content" highlights, we've had it stand out less so that the title and the summary are the key areas to look for highlighted terms.
So maybe let's do something like: 

```suggestion
    .qs-guide .content-highlights {
      font-size: 0.7rem;
      line-height: 1rem;
      color: var(--some-gray-color-variable-name);
    }
```

marko-bekhta reviewed (commented):
Looks nice! 😃

yrodiere reviewed (commented):
Thanks, looks nice indeed. I added a few comments below.

yrodiere commented on src/main/resources/web/index.html:
```diff
@@ -0,0 +1,30 @@
+    <nav class="search">
+        <div class="logo"></div>
+        <qs-form>
+            <div class="search-query">
+                <input name="q" type="search" placeholder="Start searching..."/>
+            </div>
```
Oh, and a language too.

yrodiere replied:
See https://github.com/quarkusio/search.quarkus.io/pull/210

ia3andy commented on src/test/java/io/quarkus/search/app/WebComponentsTest.java:
```diff
@@ -0,0 +1,40 @@
+
+@QuarkusTest
+@WithPlaywright
+public class WebComponentsTest {
```
Yes I need to add them

ia3andy commented on src/test/java/io/quarkus/search/app/WebComponentsTest.java:
```diff
@@ -0,0 +1,40 @@
+import io.quarkiverse.playwright.WithPlaywright;
+
+@QuarkusTest
+@WithPlaywright
```
I need to check if we can only use Chrome

ia3andy replied:
It seems it's not yet possible:
https://github.com/microsoft/playwright-java/issues/1196

The weird thing is that it should use the maven dependency and not download it AFAIK

ia3andy commented on src/main/resources/web/app/qs-form.ts:
```diff
@@ -0,0 +1,222 @@
+        if (this.localSearch) {
+            this._localSearch();
+            return;
+        }
```
I think the latest impl is working fine, let me know if you find any issue. I've tried to do something close to what you did but with the events.

ia3andy commented:
All good I think

ia3andy commented:
I think we should merge this, and iterate..

ia3andy commented:
Yeaaahh!! 🥳

ia3andy commented:
Now let me complete the quarkusio website!

yrodiere commented:
Thanks @ia3andy :)
